mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala

```diff
@@ -496,18 +499,23 @@ class ALSModel private[ml] (
               .iterator.map { j => (srcId, dstIds(j), scores(j)) }
           }
         }
-      }
-    // We'll force the IDs to be Int. Unfortunately this converts IDs to Int in the output.
-    val topKAggregator = new TopByKeyAggregator[Int, Int, Float](num, Ordering.by(_._2))
-    val recs = ratings.as[(Int, Int, Float)].groupByKey(_._1).agg(topKAggregator.toColumn)
-      .toDF("id", "recommendations")
+      }.toDF(srcOutputColumn, dstOutputColumn, ratingColumn)
+
+    val aggFunc = CollectOrdered(struct(ratingColumn, dstOutputColumn).expr, num, true)
+      .toAggregateExpression(false)
```

Comment:
I think we can define a spark sql function and wrap this part within the function, like:

```
def collect_top_k(ratingColumn, outputColumn) = {
   CollectOrdered(struct(ratingColumn, outputColumn).expr, num, true)
}
```

Reply:
sure, I think we don't want to expose it, so mark it `private[spark]`